src/Microsoft.AspNet.Routing/VirtualPathData.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+
+namespace Microsoft.AspNet.Routing
+{
+    /// <summary>
+    /// Represents information about the route and virtual path that are the result of
+    /// generating a URL with the ASP.NET routing middleware.
+    /// </summary>
+    public class VirtualPathData
+    {
+        private string _virtualPath;
+        private readonly IDictionary<string, object> _dataToken;
+
+        /// <summary>
+        ///  Initializes a new instance of the <see cref="VirtualPathData"/> class.
+        /// </summary>
+        /// <param name="router">The object that is used to generate the URL.</param>
+        /// <param name="virtualPath">The generated URL.</param>
+        public VirtualPathData([NotNull] IRouter router, string virtualPath)
+            : this(router, virtualPath, dataTokens: new RouteValueDictionary())
+        {
+        }
+
+        /// <summary>
+        ///  Initializes a new instance of the <see cref="VirtualPathData"/> class.
+        /// </summary>
+        /// <param name="router">The object that is used to generate the URL.</param>
+        /// <param name="virtualPath">The generated URL.</param>
+        /// <param name="dataTokens">The collection of custom values.</param>
+        public VirtualPathData(
+            [NotNull] IRouter router,
+            string virtualPath,
+            IDictionary<string, object> dataTokens)
+        {
+            _dataToken = dataTokens ?? new RouteValueDictionary();
```

Comment:
Got it. Actually I like moving the responsbility of taking the snapshot to the constructor. It looks simpler than before. I'll also update the following code now can be pathData = new VirtualPathData(this, tempPath, DataToken), to take advantgae of this change. Let me know if you have any other thoughts. 

// In ttps://github.com/aspnet/Routing/blob/fix154/src/Microsoft.AspNet.Routing/Template/TemplateRoute.cs#L224-L230
            pathData = new VirtualPathData(this, tempPath); 
224             if (DataTokens != null) 
225             { 
226                 foreach (var dataToken in DataTokens) 
227                 { 
228                     pathData.DataTokens.Add(dataToken.Key, dataToken.Value); 
229                 } 
230             }